### DISCUSSION TITLE
Merging pdf creating a parent bookmark for each file

### DISCUSSION THREAD

**Discussion by desplom** [py-pdf/pypdf#2797] [Q&A]
I have been investigating around and could not find an easy solution to this. I have multiple PDF, each one has its own Bookmarks. I want to merge them so as the resultant bookmarks are:

Name_of_file_1
  File1_bookmark_1
  File1_bookmark_2
  ...
Name_of_file_2
  File2_bookmark_1
  File2_bookmark_2
  ...
...

I could not find an easy way in the documentation. Is this solved anywhere? Anyone can give me some hints on how do do it?
Thanks!

**Comment by pubpub-zz** [comment]
https://pypdf.readthedocs.io/en/stable/modules/PdfWriter.html#pypdf.PdfWriter.append

outline_items allows you to create an outline/bookmark on first page and gathering all the outlines under.

**Reply by co1emi11er2** [reply]
Instead of creating an outline/bookmark on the first page, can I specify a pre-existing outline/bookmark for all the outlines of that particular file to be under?

Outline_for_file1 (separate pdf from file1)
  File1_bookmark_1
  File1_bookmark_2
...
Outline_for_file1 (separate pdf from file2)
  File2_bookmark_1
  File2_bookmark_2
...
...

**Reply by pubpub-zz** [reply]
From what I understand, you wish to modify the existing bookmarks adding a prefix. To do this I would propose you to read your files in a PdfWriter, parse, the different bookmarks and change the names :  to do that the easiest way is to directly modify the title in the raw objects through the indirect_rerence (should looks something like that `w.outline[0].indirect_reference.get_object()[NameObject("/Title"")=TextStringObject("File1_"+ w.outline[0].indirect_reference.get_object()[NameObject("/Title""))`)
and then merge the PdfWriter into a global one

**Reply by co1emi11er2** [reply]
I want to do the exact same thing as original poster, but instead of doing something like `merger.append(pdf, "Bookmark")` where "Bookmark" is the a new created bookmark on the first page and gathers all the outlines under, I would like to do `merger.append(pdf, existing_bookmark)` where `existing_bookmark` is already a bookmark in the `merger` object I am merging into and that is where all the outlines (from  `pdf` in this case) are gathered under

`merger = PdfWriter()` in this case

**Reply by pubpub-zz** [reply]
then in the individual PdfWriter before merging, delete the outlines you want to change and add new one instead. I would not recommend to try to modify directly the existing ones to the page you would like.
